### PR TITLE
KTOR-5021 Allow nested authentications to be combined using AND

### PR DESCRIPTION
**Pull request by rsinukov**

### CODE REVIEW

**Comment by e5l** [comment]
Could you please rebase on the recent changes?

**Comment by rsinukov** [comment]
@e5l done

File: ktor-server/ktor-server-plugins/ktor-server-auth/jvmAndNix/src/io/ktor/server/auth/AuthenticationInterceptors.kt [hunk 1]
```diff
@@ -63,109 +49,178 @@ private val logger = KtorSimpleLogger("Authentication")
...
+            for (provider in notRequiredProviders) {
+                if (authenticationContext._principal.principals.isNotEmpty()) break
+                if (provider.skipWhen.any { skipCondition -> skipCondition(call) }) continue
+                provider.onAuthenticate(authenticationContext)
             }
-        }
 
-        on(ChallengeHook) { call ->
-            val context = AuthenticationContext.from(call)
+            // challenge
```
Comment (e5l) [inline ktor-server/ktor-server-plugins/ktor-server-auth/jvmAndNix/src/io/ktor/server/auth/AuthenticationInterceptors.kt:87]: Please consider dropping or rephrasing

File: ktor-server/ktor-server-plugins/ktor-server-auth/jvmAndNix/src/io/ktor/server/auth/AuthenticationInterceptors.kt [hunk 2]
```diff
@@ -63,109 +49,178 @@ private val logger = KtorSimpleLogger("Authentication")
...
-            for (provider in configurations) {
+            if (authenticationContext.principal<Principal>() != null) return@on
+
+            var count = 0
+            for (provider in requiredProviders) {
                 if (provider.skipWhen.any { skipCondition -> skipCondition(call) }) continue
 
                 provider.onAuthenticate(authenticationContext)
+                count++
+                if (authenticationContext._principal.principals.size < count) {
```
Comment (e5l) [inline ktor-server/ktor-server-plugins/ktor-server-auth/jvmAndNix/src/io/ktor/server/auth/AuthenticationInterceptors.kt:75]: What if the different(some optional) principals are saved and required are not? Could you add a test?

File: ktor-server/ktor-server-plugins/ktor-server-auth/jvmAndNix/src/io/ktor/server/auth/AuthenticationInterceptors.kt [hunk 3]
```diff
@@ -63,109 +49,178 @@ private val logger = KtorSimpleLogger("Authentication")
...
-                    }
-                    return@on
-                }
+private suspend fun executeChallenges(context: AuthenticationContext, call: ApplicationCall) {
+    val challenges = context.challenge.challenges
+
+    for (challenge in challenges) {
+        challenge(context.challenge, call)
+        if (context.challenge.completed) {
+            if (!call.isHandled) {
```
Comment (e5l) [inline ktor-server/ktor-server-plugins/ktor-server-auth/jvmAndNix/src/io/ktor/server/auth/AuthenticationInterceptors.kt:107]: braces can be dropped to make it a bit more readable

File: ktor-server/ktor-server-plugins/ktor-server-auth/jvmAndNix/src/io/ktor/server/auth/AuthenticationInterceptors.kt [hunk 4]
```diff
@@ -63,109 +49,178 @@ private val logger = KtorSimpleLogger("Authentication")
...
-                    if (!call.isHandled) {
-                        call.respond(UnauthorizedResponse())
-                    }
-                    return@on
-                }
+private suspend fun executeChallenges(context: AuthenticationContext, call: ApplicationCall) {
+    val challenges = context.challenge.challenges
+
+    for (challenge in challenges) {
+        challenge(context.challenge, call)
```
Comment (e5l) [inline ktor-server/ktor-server-plugins/ktor-server-auth/jvmAndNix/src/io/ktor/server/auth/AuthenticationInterceptors.kt:105]: could we extract this block from several loops to a single method?

File: ktor-server/ktor-server-plugins/ktor-server-auth/jvmAndNix/src/io/ktor/server/auth/AuthenticationInterceptors.kt [hunk 5]
```diff
@@ -63,109 +49,178 @@ private val logger = KtorSimpleLogger("Authentication")
...
-                    logger.trace("Responding unauthorized because of error ${error.message}")
-                    if (!call.isHandled) {
-                        call.respond(UnauthorizedResponse())
-                    }
-                    context.challenge.complete()
-                    return@on
-                }
+    for (error in context.allErrors) {
+        if (!context.challenge.completed) {
+            logger.trace("Responding unauthorized because of error ${error.message}")
```
Comment (e5l) [inline ktor-server/ktor-server-plugins/ktor-server-auth/jvmAndNix/src/io/ktor/server/auth/AuthenticationInterceptors.kt:126]: please consider adding `call.request.uri` here

Comment (e5l) [reply ktor-server/ktor-server-plugins/ktor-server-auth/jvmAndNix/src/io/ktor/server/auth/AuthenticationInterceptors.kt:126]: we should also log in other cases

**Review by e5l — CHANGES_REQUESTED** [review]
please address some comments, LGTM in general

**Comment by e5l** [comment]
@andreyaksenov, could you please check KDocs?

File: ktor-server/ktor-server-plugins/ktor-server-auth/jvmAndNix/src/io/ktor/server/auth/AuthenticationInterceptors.kt (hunk 2 above)
Comment (rsinukov) [reply ktor-server/ktor-server-plugins/ktor-server-auth/jvmAndNix/src/io/ktor/server/auth/AuthenticationInterceptors.kt:75]: These block checks only required providers. There are tests for different combinations of strategies in this PR.

File: ktor-server/ktor-server-plugins/ktor-server-auth/jvmAndNix/src/io/ktor/server/auth/AuthenticationInterceptors.kt (hunk 5 above)
Comment (rsinukov) [reply ktor-server/ktor-server-plugins/ktor-server-auth/jvmAndNix/src/io/ktor/server/auth/AuthenticationInterceptors.kt:126]: I'm waiting for your PR with logging to be merged to update mine after

File: ktor-server/ktor-server-plugins/ktor-server-auth/jvmAndNix/src/io/ktor/server/auth/AuthenticationInterceptors.kt [hunk 6]
```diff
@@ -63,109 +49,178 @@ private val logger = KtorSimpleLogger("Authentication")
...
+            "Authentication configuration with the name $configurationName was not found"
+        }
+    )
+}
+
+/**
+ *  A resolution strategy for nested authentication providers.
+ *  [AuthenticationStrategy.Optional] - if no authentication is provided by the client,
+ *  a call continues but with a null [Principal].
+ *  [AuthenticationStrategy.FirstSuccessful] - client must provide authentication for at least one provider
```
Comment (andreyaksenov) [inline ktor-server/ktor-server-plugins/ktor-server-auth/jvmAndNix/src/io/ktor/server/auth/AuthenticationInterceptors.kt:159]: `must provide authentication` -> `must provide authentication credentials` or `must provide authentication data`
Please update here and in other places as well

**Review by andreyaksenov — CHANGES_REQUESTED** [review]
looks good, only one comment

**Review by andreyaksenov — APPROVED** [review]

File: ktor-server/ktor-server-plugins/ktor-server-auth/jvmAndNix/src/io/ktor/server/auth/AuthenticationInterceptors.kt [hunk 7]
```diff
@@ -65,22 +51,47 @@ public val AuthenticationInterceptors: RouteScopedPlugin<RouteAuthenticationConf
...
+            }
+
+            LOGGER.trace("Trying to authenticate ${call.request.uri} with required ${provider.name}")
+            provider.onAuthenticate(authenticationContext)
+            count++
+            if (authenticationContext._principal.principals.size < count) {
+                LOGGER.trace("Authentication failed for ${call.request.uri} with provider $provider")
+                authenticationContext.executeChallenges(call)
+                return@on
+            } else {
```
Comment (e5l) [inline ktor-server/ktor-server-plugins/ktor-server-auth/jvmAndNix/src/io/ktor/server/auth/AuthenticationInterceptors.kt:85]: else is not required

**Review by e5l — APPROVED** [review]
lgtm

**Pull request merged**